x509/src/certificate.rs

```diff
@@ -21,9 +24,65 @@ pub fn default_zero() -> u32 {
     0
 }
 
-/// only support v3 certificates
+/// only supporting v3 certificates, but all versions are listed here
 /// Version  ::=  INTEGER  {  v1(0), v2(1), v3(2)  }
-pub const X509_CERT_VERSION: u8 = 2;
+
+/// Version identifier for X.509 certificates. In practice, only v3 is used.
+#[derive(Clone, Debug, Copy, PartialEq, Eq, PartialOrd)]
+pub enum Version {
+    /// Denotes X.509 v1
+    V1 = 0,
+
+    /// Denotes X.509 v2 - added issuerUniqueID and subjectUniqueID (both of which have been deprecated)
+    V2 = 1,
+
+    /// Denotes X.509 v3 - add extensions
+    V3 = 2,
+}
+
+impl Default for Version {
+    fn default() -> Self {
+        Version::V3
+    }
+}
+
+impl Decodable<'_> for Version {
+    fn decode(decoder: &mut Decoder<'_>) -> der::Result<Self> {
+        Version::try_from(u8::decode(decoder)?).map_err(|_| Self::TAG.value_error())
+    }
+}
+
+impl Encodable for Version {
+    fn encoded_len(&self) -> der::Result<der::Length> {
+        Length::ONE.for_tlv()
+    }
+
+    fn encode(&self, encoder: &mut Encoder<'_>) -> der::Result<()> {
+        u8::from(*self).encode(encoder)
+    }
+}
```

Comment:
Why not `DecodeValue` and `EncodeValue` here?

Reply:
No particular reason, habit I suppose.

Reply:
Actually, Decodable/Encodable were taken from PKCS8 Version

Reply:
Aah yeah, that's outdated, and also PKCS#8 doesn't need to deal with a context-specific version field so it works there